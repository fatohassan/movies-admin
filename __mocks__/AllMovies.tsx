import { MockedProvider, MockedResponse } from "@apollo/client/testing";
import { renderHook } from "@testing-library/react";
import React from "react";
import  MoviesList  from "../src/pages/admin/movies/MoviesList";
import { LOAD_MOVIES } from "../src/pages/admin/movieQueries/moviesQueries";

export const AllMovies = {
    getMoviesList: [
    {
      id: "66715c4f88c42067190e070f",
      original_title: "Inside Out 2",
      release_date: "2024-06-11",
      vote_average: "7.761",
      backdrop_path: "/coATv42PoiLqAFKStJiMZs2r6Zb.jpg",
      image: null,
      overview:
        "Teenager Riley's mind headquarters is undergoing a sudden demolition to make room for something entirely unexpected: new Emotions! Joy, Sadness, Anger, Fear and Disgust, who’ve long been running a successful operation by all accounts, aren’t sure how to feel when Anxiety shows up. And it looks like she’s not alone.",
    },
  ],
};

// successful mock of getMovieList
export const successfulMoviesMock: MockedResponse[]= [
    {
        request: {
            query: LOAD_MOVIES
        },
        result: {
            data: AllMovies
        }
    }
] 

export function getAllMoviesWrapper(mockData: MockedResponse[] = []) {
    const wrapper =({children}: React.PropsWithChildren) => (
        <MockedProvider mocks={mockData} addTypename={false}>
            {children}
        </MockedProvider>
    );
    const {result}:any = renderHook(() => MoviesList(), {wrapper})
    return {
        result
    }
}
