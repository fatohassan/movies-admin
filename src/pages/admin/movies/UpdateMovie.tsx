import axios from "axios";
import { useEffect, useState } from "react";
// import { useState } from "react";
import { Link, useNavigate, useParams } from "react-router-dom";

export default function UpdateMovie() {
  // const [validationErrors, setValidationErrors] = useState({})
  const params = useParams();
  const [initialState, setInitialState] = useState();
  const navigate = useNavigate();

  const getMovies = async () => {
    try {
      const response = await axios.get("http://localhost:5000/movie/" + params.id);
      const data = await response.data;
      setInitialState(data);

      if (response.status === 201) {
        navigate("/admin/movies/");
      } 
      // else if (response.status === 400) {
      //   alert("validation error");
      // } else {
      //   alert("unable to read movie");
      // }
    } catch (error) {
      console.log(error);
      alert("unable to connect to server");
    }
  };

  useEffect(() => {getMovies()}, [])

  const handleSubmit = async (event: React.FormEvent<HTMLFormElement>) => {
    event.preventDefault();

    const formData = new FormData(event.currentTarget);
    const movie = Object.fromEntries(formData.entries());
    try {
      const response = await axios.put(
        "http://localhost:5000/movie/"+ params.id,
        formData
      );
      const data = await response.data;

      if (response.status === 201) {
        navigate("/admin/movies/");
      } else if (response.status === 400) {
        alert("validation error");
      } else {
        alert("unable to create movie");
      }
    } catch (error) {
      console.log(error);
      alert("unable to connect to server");
    }
  };

  type elements = {
    id: any;
    original_title: any;
  };
  return (
    <div className="container my-4">
      <div className="row">
        <div className="col-md-8 mx-auto rounded border p-4">
          <h2 className="text-center mb-5">Edit Product</h2>

          <div className="row mb-3">
            <label className="col-sm-4 col-form-label">ID</label>
            <div className="col-sm-8">
              <input
                readOnly
                className="form-control-plaintext"
                defaultValue={params.id}
              />
            </div>
          </div>
          {initialState && (
            <form onSubmit={handleSubmit}>
              {/* <div className="row mb-3">
              <label className="col-sm-4 col-form-label">ID</label>
              <div className="col-sm-8">
                <input className="form-control" name="id" />
                <span className="text-danger"></span>
              </div>
            </div> */}
              <div className="row mb-3">
                <label className="col-sm-4 col-form-label">Title</label>
                <div className="col-sm-8">
                  <input className="form-control" name="original_title" 
                  // defaultValue={initialState.original_title}
                  />
                  <span className="text-danger"></span>
                </div>
              </div>
              <div className="row mb-3">
                <label className="col-sm-4 col-form-label">Rating</label>
                <div className="col-sm-8">
                  <input className="form-control" name="rating" type="number" 
                  // defaultValue={initialState.vote_average}
                  />
                  <span className="text-danger"></span>
                </div>
              </div>
              <div className="row mb-3">
                <label className="col-sm-4 col-form-label">Release Date</label>
                <div className="col-sm-8">
                  <input className="form-control" name="release_date" 
                  // defaultValue={initialState.release_date}
                  />
                  <span className="text-danger"></span>
                </div>
              </div>
              <div className="row mb-3">
                <label className="col-sm-4 col-form-label">Description</label>
                <div className="col-sm-8">
                  <textarea className="form-control" name="overview" rows={3} 
                  // defaultValue={initialState.overview}
                  />
                  <span className="text-danger"></span>
                </div>
              </div>
              {/* <div className="row mb-3">
              <div className="offset-sm-4 col-sm-8">
                <img src='' width='150' alt='..'/>
              </div>
            </div> */}
              <div className="row mb-3">
                <label className="col-sm-4 col-form-label">Image</label>
                <div className="col-sm-8">
                  <input className="form-control" name="image" type="file" 
                  // defaultValue={initialState.backdrop_path}
                  />
                  <span className="text-danger"></span>
                </div>
              </div>

              <div className="row">
                <div className="offset-sm-4 col-sm-4 d-grid">
                  <button type="submit" className="btn btn-primary">
                    Submit
                  </button>
                </div>
                <div className="col-sm-4 d-grid">
                  <Link
                    className="btn btn-secondary"
                    to="/admin/movies"
                    role="button"
                  >
                    Cancel
                  </Link>
                </div>
              </div>
            </form>
          )}
        </div>
      </div>
    </div>
  );
}
