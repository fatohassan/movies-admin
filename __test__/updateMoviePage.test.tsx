import { cleanup, fireEvent, render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import UpdateMovie from "../src/pages/admin/movies/UpdateMovie";
import React from "react";
import { MockedProvider } from "@apollo/client/testing";
import renderer from "react-test-renderer";
import { BrowserRouter } from "react-router-dom";
import { updateMovieMutationsMock } from "../__mocks__/AllMovies";

const renderUpdateMovie = () => {
  return (
    <BrowserRouter>
      <MockedProvider mocks={updateMovieMutationsMock} addTypename={false}>
        <UpdateMovie />
      </MockedProvider>
    </BrowserRouter>
  );
};
describe("Update Movie rendering navigating", () => {
  it("should render page correctly", () => {
    const tree = renderer
      .create(
        <BrowserRouter>
          <MockedProvider mocks={updateMovieMutationsMock} addTypename={false}>
            <UpdateMovie />
          </MockedProvider>
        </BrowserRouter>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should display contents", () => {
    render(renderUpdateMovie());
    const headingElement = screen.getByRole("heading", {
      name: /Edit Movie/i,
    });
    expect(headingElement).toBeInTheDocument();
    expect(screen.getByText("ID")).toBeVisible();
  });

  it("should display form contents/inputs/buttons", async () => {
    render(renderUpdateMovie());

    screen.getByLabelText("form");
    await screen.findByText("Title");
    expect(await screen.findByText("Title")).toBeInTheDocument();
    expect(screen.getByRole("button", { name: "Submit" })).toBeTruthy();
    expect(screen.getByLabelText("link")).toHaveTextContent("Cancel");
  });

  it("should be able to type in title inuput", () => {
    render(renderUpdateMovie());
    const inputTitle = screen.getByLabelText("title");
    expect(
      fireEvent.change(inputTitle, { EventTarget: { value: "Inside Out 2" } })
    ).toBeTruthy();
  });

  it("should be able to type in rating inuput", () => {
    render(renderUpdateMovie());
    const inputRating = screen.getByLabelText("rating");
    expect(
      fireEvent.change(inputRating, { EventTarget: { value: "1" } })
    ).toBeTruthy();
  });

  it("should be able to type in description inuput", () => {
    render(renderUpdateMovie());
    const inputOverview = screen.getByLabelText("overview");
    fireEvent.change(inputOverview, { EventTarget: { value: "Good Movie" } });
    expect(inputOverview).toBeVisible();
  });

  it("should be able to type in description inuput", () => {
    render(renderUpdateMovie());
    const inputReleaseDate = screen.getByLabelText("date");
    expect(
      fireEvent.change(inputReleaseDate, {
        EventTarget: { value: "01-02-2024" },
      })
    ).not.toBeFalsy();
  });

  it("should be able to type in image inuput", () => {
    render(renderUpdateMovie());
    const inputImage = screen.getByLabelText("image");
    expect(inputImage).toBeInTheDocument();
  });
});
