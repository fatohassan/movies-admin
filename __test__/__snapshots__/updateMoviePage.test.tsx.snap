// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Update Movie rendering navigating should render page correctly 1`] = `
<div
  className="container my-4"
>
  <div
    className="row"
  >
    <div
      className="col-md-8 mx-auto rounded border p-4"
    >
      <h2
        className="text-center mb-5"
      >
        Edit Movie
      </h2>
      <div
        className="row mb-3"
      >
        <label
          className="col-sm-4 col-form-label"
        >
          ID
        </label>
        <div
          className="col-sm-8"
        >
          <input
            className="form-control-plaintext"
            readOnly={true}
          />
        </div>
      </div>
      <form
        aria-label="form"
        onSubmit={[Function]}
      >
        <div
          className="row mb-3"
        >
          <label
            className="col-sm-4 col-form-label"
          >
            Title
          </label>
          <div
            className="col-sm-8"
          >
            <input
              aria-label="title"
              className="form-control"
              id="original_title"
              name="original_title"
            />
            <span
              className="text-danger"
            />
          </div>
        </div>
        <div
          className="row mb-3"
        >
          <label
            className="col-sm-4 col-form-label"
          >
            Rating
          </label>
          <div
            className="col-sm-8"
          >
            <input
              aria-label="rating"
              className="form-control"
              name="vote_average"
              type="number"
            />
            <span
              className="text-danger"
            />
          </div>
        </div>
        <div
          className="row mb-3"
        >
          <label
            className="col-sm-4 col-form-label"
          >
            Release Date
          </label>
          <div
            className="col-sm-8"
          >
            <input
              aria-label="date"
              className="form-control"
              name="release_date"
            />
            <span
              className="text-danger"
            />
          </div>
        </div>
        <div
          className="row mb-3"
        >
          <label
            className="col-sm-4 col-form-label"
          >
            Description
          </label>
          <div
            className="col-sm-8"
          >
            <textarea
              aria-label="overview"
              className="form-control"
              name="overview"
              rows={3}
            />
            <span
              className="text-danger"
            />
          </div>
        </div>
        <div
          className="row mb-3"
        >
          <label
            className="col-sm-4 col-form-label"
          >
            Image
          </label>
          <div
            className="col-sm-8"
          >
            <img
              src="https://image.tmdb.org/t/p/w1280undefined"
              width="200"
            />
            <input
              aria-label="image"
              className="form-control mt-4"
              name="image"
              type="file"
            />
            <span
              className="text-danger"
            />
          </div>
        </div>
        <div
          className="row"
        >
          <div
            className="offset-sm-4 col-sm-4 d-grid"
          >
            <button
              className="btn btn-primary"
              type="submit"
            >
              Submit
            </button>
          </div>
          <div
            className="col-sm-4 d-grid"
          >
            <a
              aria-label="link"
              className="btn btn-secondary"
              href="/admin/movies"
              onClick={[Function]}
              role="button"
            >
              Cancel
            </a>
          </div>
        </div>
      </form>
    </div>
  </div>
</div>
`;
